server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/aggregators/SumValueAggregator.java

```diff
@@ -21,87 +22,107 @@
-
     public SumValueAggregator(FieldValueConverter fieldValueConverter) {
         this.fieldValueConverter = fieldValueConverter;
+        this.compensatedSumConverter = new CompensatedSumType();
     }
 
     @Override
     public FieldValueConverter getAggregatedValueType() {
-        return VALUE_AGGREGATOR_TYPE;
+        return compensatedSumConverter;
     }
 
     @Override
-    public Double getInitialAggregatedValueForSegmentDocValue(Long segmentDocValue) {
-        kahanSummation.reset(0, 0);
+    public CompensatedSum getInitialAggregatedValueForSegmentDocValue(Long segmentDocValue) {
+        CompensatedSum kahanSummation = new CompensatedSum(0, 0);
         // add takes care of the sum and compensation internally
         if (segmentDocValue != null) {
-            kahanSummation.add(fieldValueConverter.toDoubleValue(segmentDocValue));
+            kahanSummation.reset(fieldValueConverter.toDoubleValue(segmentDocValue), 0);
         } else {
-            kahanSummation.add(getIdentityMetricValue());
+            kahanSummation.reset(getIdentityMetricDoubleValue(), 0);
         }
-        return kahanSummation.value();
+        return kahanSummation;
     }
 
     // we have overridden this method because the reset with sum and compensation helps us keep
     // track of precision and avoids a potential loss in accuracy of sums.
     @Override
-    public Double mergeAggregatedValueAndSegmentValue(Double value, Long segmentDocValue) {
-        assert value == null || kahanSummation.value() == value;
+    public CompensatedSum mergeAggregatedValueAndSegmentValue(CompensatedSum value, Long segmentDocValue) {
+        CompensatedSum kahanSummation = new CompensatedSum(0, 0);
+        if (value != null) {
+            kahanSummation.reset(value.value(), value.delta());
+        }
         // add takes care of the sum and compensation internally
         if (segmentDocValue != null) {
             kahanSummation.add(fieldValueConverter.toDoubleValue(segmentDocValue));
         } else {
-            kahanSummation.add(getIdentityMetricValue());
+            kahanSummation.add(getIdentityMetricDoubleValue());
         }
-        return kahanSummation.value();
+        return kahanSummation;
     }
 
     @Override
-    public Double mergeAggregatedValues(Double value, Double aggregatedValue) {
-        assert aggregatedValue == null || kahanSummation.value() == aggregatedValue;
-        // add takes care of the sum and compensation internally
+    public CompensatedSum mergeAggregatedValues(CompensatedSum value, CompensatedSum aggregatedValue) {
+        CompensatedSum kahanSummation = new CompensatedSum(0, 0);
+        if (aggregatedValue != null) {
+            kahanSummation.reset(aggregatedValue.value(), aggregatedValue.delta());
+        }
         if (value != null) {
-            kahanSummation.add(value);
+            kahanSummation.add(value.value(), value.delta());
```

Comment:
For the precision difference you mentioned, can it be due to now relying on delta of the first param as well?